cmd/multifile.go

```diff
@@ -0,0 +1,182 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"net/url"
+	"os"
+
+	"github.com/replicate/pget/pkg/config"
+	"github.com/replicate/pget/pkg/download"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/replicate/pget/pkg/optname"
+)
+
+const multiFileLongDesc = `
+'multifile' mode for pget takes a manifest file as input (can use '-' for stdin) and downloads all files listed in the manifest.
+
+The manifest is expected to be in the format of a newline-separated list of pairs of URLs and destination paths, separated by a space.
+e.g.
+https://example.com/file1.txt /tmp/file1.txt
+
+'multifile'' will download files in parallel limited to the '--maximum-connections-per-host' limit for per-host limts and 
+over-all limited to the '--max-concurrency' limit for overall concurrency.
+`
+
+// multifile mode config vars
+var (
+	MultifileMaxConnPerHost            int
+	MultifileSerializePerHostDownloads bool
+)
+
+var MultiFileCMD = &cobra.Command{
+	Use:   "multifile [flags] <manifest-file>",
+	Short: "download files from a manifest file in parallel",
+	Long:  multiFileLongDesc,
+	Args:  cobra.ExactArgs(1),
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		if viper.GetInt(optname.MaxConnPerHost) == 0 {
+			viper.Set(optname.MaxConnPerHost, 40)
+		}
+		if viper.GetBool(optname.Extract) {
+			return fmt.Errorf("cannot use --extract with multifile mode")
+		}
+		return nil
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		if err := execMultifile(cmd, args); err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+	},
+}
+
+func init() {
+	RootCMD.AddCommand(MultiFileCMD)
+	MultiFileCMD.PersistentFlags().IntVar(&MultifileMaxConnPerHost, optname.MaxConnPerHost, 0, "Maximum number of connections per host")
+	MultiFileCMD.PersistentFlags().BoolVar(&MultifileSerializePerHostDownloads, optname.SerializePerHostDownloads, true, "Whether to serialize downloads from the same host")
+	err := viper.BindPFlags(MultiFileCMD.PersistentFlags())
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+type manifestEntry struct {
+	url  string
+	dest string
+}
+
+func execMultifile(cmd *cobra.Command, args []string) error {
+	var reader *bufio.Reader
+	// if manifest file is '-', read from stdin
+	manifestPath := args[0]
+	if manifestPath == "-" {
+		reader = bufio.NewReader(os.Stdin)
+	} else {
+		// check that the manifest file exists
+		_, err := os.Stat(manifestPath)
+		if os.IsNotExist(err) {
+			return fmt.Errorf("manifest file %s does not exist", manifestPath)
+		}
+		manifestFile, err := os.Open(manifestPath)
+		if err != nil {
+			return fmt.Errorf("error opening manifest file %s: %w", manifestPath, err)
+		}
+		defer manifestFile.Close()
+		reader = bufio.NewReader(manifestFile)
+	}
+	// process the manifest file into a map of hosts to url/file pairs with processManifest
+	buffer := make([]string, 0)
+	for {
+		readString, err := reader.ReadString('\n')
```

Comment:
we could also use a bufio.Scanner here https://pkg.go.dev/bufio#Scanner

Reply:
though I think this would be simpler if we did os.ReadFile and just operated on the whole manifest file in memory https://pkg.go.dev/os#ReadFile

Reply:
Yeah Readfile is better. Will fix

Reply:
Oh! We can't use ReadFile, since we need to rely on stdin as well, we can use readall though